Add 'g:lightline#gitdiff#show_empty_indicators' option

gblock0 opened the pull request:
I like being able to see `A: 0 D: 0 M: 0`, so this adds the `g:lightline#gitdiff#show_empty_indicators` option to show 0 for the indicators if they would otherwise be empty.

This is my first Vim script related PR, so let me know if you'd prefer things any changes.

niklaas commented:
Thanks a lot for your contribution. I'll take a look at it.

niklaas commented on autoload/lightline/gitdiff/algorithms/word_diff_porcelain.vim:
```diff
@@ -17,16 +17,17 @@ function! lightline#gitdiff#algorithms#word_diff_porcelain#calculate(buffer) abo
   let l:lines_modified = len(filter(copy(l:changes), { idx, val -> val ==# 'M' }))
 
   let l:ret = {}
+  let l:show_empty_indicators = exists('g:lightline#gitdiff#show_empty_indicators') && g:lightline#gitdiff#show_empty_indicators
```
You could use the following, which does checking for existence of the key. If the key doesn't exist, zero is returned, see `:h get()`.

```suggestion
  let l:show_empty_indicators = get('g:lightline#gitdiff#show_empty_indicators')
```

niklaas commented on README.md:
````diff
@@ -115,6 +115,11 @@ provide your own. Take a look at the source of both functions for inspiration
 feature-rich algorithms in the package.
 
+You can show empty indicators (i.e. `A: 0 D: 0 M: 0`) in the following way:
+```vim
````
Just for keeping the formatting style.

```suggestion

```vim
```

gblock0 commented:
Thanks for the feedback! I’ll take a look tonight and make the changes.

gblock0 commented:
I'll add tests for `show_empty_indicators` later tonight

gblock0 commented:
Tests added and code updated! Let me know your thoughts!

These were my first Vader tests, so let me know if you'd like them structured differently.

niklaas commented on test/lightline-gitdiff.vader:
```diff
@@ -1,2 +1,22 @@
 Include (Utils): utils/group_at.vader
+
+Execute(write_calculation_to_cache(): given no show_empty_indicators variable):
+  call g:lightline#gitdiff#write_calculation_to_cache(1, 0)
```
Thanks a lot for providing the tests. I haven't looked at the code for a while and had to find out the syntax for Vader.vim and understand structure of the test files I wrote originally myself.

To write these tests in a clean manner, we should mock the used algorithm. At the current state, calculation is done on buffer 1 -- to be honest, I don't really know what checked there. The following is a sketch for mocking the algorithm and providing a test for mixed results (such as `A: 1 M: 0 D: 1`); we would need to apply this to the subsequent tests as well:

```diff
diff --git a/test/lightline-gitdiff.vader b/test/lightline-gitdiff.vader
index fc926ff..48c60cf 100644
--- a/test/lightline-gitdiff.vader
+++ b/test/lightline-gitdiff.vader
@@ -1,12 +1,27 @@
 Include (Algorithms): algorithm/parse_indicator_group.vader
 Include (Utils): utils/group_at.vader
 
-Execute(write_calculation_to_cache(): given no show_empty_indicators variable):
+Before :
+  if exists('g:LightlineAlgorithm')
+    unlet g:LightlineAlgorithm
+  endif
+
+Execute(write_calculation_to_cache(): given no show_empty_indicators variable with empty result):
+  let g:LightlineAlgorithm = { -> { 'foo': 0, 'bar': 0 } }
+
   call g:lightline#gitdiff#write_calculation_to_cache(1, 0)
   let actual = get(g:, 'lightline#gitdiff#cache')[1]
 Then (should return no empty indicators):
   AssertEqual {}, actual
 
+Execute(write_calculation_to_cache(): given no show_empty_indicators variable with mixed result):
+  let g:LightlineAlgorithm = { -> { 'foo': 1, 'bar': 0 } }
+
+  call g:lightline#gitdiff#write_calculation_to_cache(1, 0)
+  let actual = get(g:, 'lightline#gitdiff#cache')[1]
+Then (should remove the empty key only):
+  AssertEqual { 'foo': 1 }, actual
+
 Execute(write_calculation_to_cache(): given g:lightline#gitdiff#show_empty_indicators == 0):
   let g:lightline#gitdiff#show_empty_indicators = 0
   call lightline#gitdiff#write_calculation_to_cache(2, 0)
```

Note the new variable `g:LightlineAlgorithm`. While writing this, I figured there is a bug in the plugin. On the contrary to what's written in den readme, it's not possible to provide a custom algorithm because, in vimscript, global functions must start with a capital letter. Because of this, I also suggest the following changes:

```diff
diff --git a/README.md b/README.md
index ab17561..edf0511 100644
--- a/README.md
+++ b/README.md
@@ -106,7 +106,7 @@ By default, the latter one is used because it allows to display modified lines.
 This resembles the default:
 
 ```vim
-let g:lightline#gitdiff#algorithm =
+let g:LightlineAlgorithm =
       \ { buffer -> lightline#gitdiff#algorithms#word_diff_porcelain#calculate(buffer) }
 
diff --git a/autoload/lightline/gitdiff.vim b/autoload/lightline/gitdiff.vim
index c1ec1c3..c0d0b3b 100644
--- a/autoload/lightline/gitdiff.vim
+++ b/autoload/lightline/gitdiff.vim
@@ -37,7 +37,7 @@ function! lightline#gitdiff#write_calculation_to_cache(buffer, soft) abort
     return
   endif
 
-  let l:indicator_values = get(g:, 'lightline#gitdiff#algorithm',
+  let l:indicator_values = get(g:, 'LightlineAlgorithm',
       \ { buffer -> lightline#gitdiff#algorithms#word_diff_porcelain#calculate(buffer) })(a:buffer)
 
   " If the user doesn't want to show empty indicators,
@@ -73,7 +73,7 @@ endfunction
 "
 " In fact, an arbitrary number of changes can be supported. This depends on
 " the algorithm that is used for calculation
-" (`g:lightline#gitdiff#algorithm`). However, this function takes only these
+" (`g:LightlineAlgorithm`). However, this function takes only these
 " types of changes into account b/c it only provides default indicators for
 " these types. If an algorithm does not support a particular type, this is not
 " an issue; if it supports more types than this function, the additional types
```

gblock0 commented on autoload/lightline/gitdiff.vim:
```diff
@@ -34,12 +34,23 @@ endfunction
+
+  " If the user doesn't want to show empty indicators,
+  "     then remove the empty indicators returned from the algorithm
+  if !get(g:, 'lightline#gitdiff#show_empty_indicators', 0)
+    for key in keys(l:indicator_values)
+      if l:indicator_values[key] == 0
+        unlet l:indicator_values[key]
+      endif
+    endfor
+  endif
```
Wouldn't this keep the indicators if the user hadn't set the `show_empty_indicators` variable (or set it to `0`) instead of removing them?

My thought was to have the algorithm return the "full" git state and have `write_calculation_to_cache` figure out which indicators to show.

niklaas replied:
This is `write_calculation_to_cache()` isn't it? 🙂

Anyway, you're totally right, I mixed up the boolean. But the version you provided should be more perfomant because checking for the configuration value is only done once.

gblock0 commented on autoload/lightline/gitdiff/algorithms/word_diff_porcelain.vim:
```diff
@@ -16,21 +16,7 @@ function! lightline#gitdiff#algorithms#word_diff_porcelain#calculate(buffer) abo
-  endif
-
-  return l:ret
+  return { 'A': l:lines_added, 'D': l:lines_deleted, 'M': l:lines_modified}
```
See my comment above about why I returned the whole object.

gblock0 commented:
No worries at all! I really appreciate all the great feedback! 

- I added a lot of tests based on your feedback. I think I covered all the permutations, but let me know if I missed any
- I renamed `lightline#gitdiff#algorithm` to `LightlineGitDiffAlgorithm` (I updated the README and comments to reflect this as well)
- `numstat.vim` now always returns an object with `A` and `D`
- I kept the change in `word_diff_porcelain.vim` the same to let `write_calculation_to_cache` process which indicators should be shown, but happy to change it if you still want me to


I've checked `Allow edits by maintainers` if you'd like to make any more changes, but I'm also happy to make any other necessary changes!

niklaas commented:
I really appreciate your contributions and thoughts. I’ll take a look at these throughout the day.

niklaas commented:
P.S.: Since the plugin seems to get a bit more attention, what do you think of #16 ? I'm wondering whether I should spend some time on that.

niklaas commented:
🎉 thanks again!

gblock0 commented:
Happy to contribute! I'll take a look at #16 and add some thoughts.
